test/test_linearizer.py

```diff
@@ -210,10 +210,14 @@ def test_multireduce_loop_scope(self):
     output_shape = tuple([1 if i == axis else x for i,x in enumerate(list(shape))])
     load = MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker.from_shape(shape))
     store = MemBuffer(idx=0, dtype=dtypes.float, st=ShapeTracker.from_shape(output_shape))
+    const = ConstBuffer(val=1/64, dtype=dtypes.float, st=ShapeTracker.from_shape((1,1,1)))
     ast = LazyOp(op=BufferOps.STORE, src=(LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BinaryOps.SUB, src=(
-      LazyOp(op=BufferOps.LOAD, src=(), arg=load), LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load),)
-      , arg=((axis,), dtypes.float))), arg=None),), arg=((axis,), dtypes.float)),), arg=store)
+      LazyOp(op=BufferOps.LOAD, src=(), arg=load), LazyOp(op=BinaryOps.MUL, src=(
+        LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load),), arg=((axis,), dtypes.float)),
+        LazyOp(op=BufferOps.CONST, src=(), arg=const),
+        )), ), arg=None),), arg=((axis,), dtypes.float)),), arg=store)
     k = Linearizer(ast)
+    k.hand_coded_optimizations()
```

Comment:
this will create the `DEFINE_LOCAL` that causes problems with loop_scope when rendering multiple reduceops

b/c the `MUL` uses a previous `SUM` as an input it won't get moved around by `fix_loop_scope`, this isn't necessarily problematic in this case it will just do the calculation for every iteration of the loop. BUT if the result of the `MUL` gets used later, outside this loop, it will cause compilation issues because of scope

Reply:
Is this fixed?

Reply:
the test? it ought to be

there is something strange where in the rendered kernel it will move the `MUL` back into the loop if it's only used within the loop scope despite it coming before the `RANGE` in the uops

if the MUL is not used outside the loop
*kernel:*
```
...
  barrier(CLK_LOCAL_MEM_FENCE);
  float val2 = temp[0];
  float acc2 = 0.0f;
  for (int ridx2 = 0; ridx2 < 2; ridx2++) {
    float val3 = data1[alu1+ridx2];
    float alu2 = (val3-(val2*0.03125f));
    acc2 = ((alu2*alu2)+acc2);
  }
  temp[lidx1] = acc2;
  barrier(CLK_LOCAL_MEM_FENCE);
...
```
*uops:*
```
...
  33 UOps.BARRIER        :                           [31]                             None
  34 UOps.LOAD           : dtypes.float              [2, 3, 33]                       None
  35 UOps.ALU            : dtypes.float              [34, 4]                          BinaryOps.MUL
  36 UOps.DEFINE_ACC     : dtypes.float              [37]                             (0.0, '1_-1', 0)
  37 UOps.RANGE          : dtypes.int                [3, 6]                           (6, 0)
  38 UOps.ALU            : dtypes.int                [14, 37]                         BinaryOps.ADD
  39 UOps.LOAD           : dtypes.float              [1, 38]                          None
  40 UOps.ALU            : dtypes.float              [39, 35]                         BinaryOps.SUB
  41 UOps.ALU            : dtypes.float              [40, 40]                         BinaryOps.MUL
  42 UOps.ALU            : dtypes.float              [41, 36]                         BinaryOps.ADD
  43 UOps.PHI            : dtypes.float              [36, 42]                         None
  44 UOps.ENDRANGE       :                           [37]                             None
  45 UOps.STORE          :                           [2, 10, 43]                      None
  46 UOps.BARRIER        :                           [45]                             None
...
```

versus if the MUL is used outside the loop
*kernel*
```
...
  barrier(CLK_LOCAL_MEM_FENCE);
  float val2 = temp[0];
  float alu2 = (val2*0.03125f);
  float acc2 = 0.0f;
  for (int ridx2 = 0; ridx2 < 2; ridx2++) {
    float val3 = data1[alu1+ridx2];
    float alu3 = (val3-alu2);
    acc2 = ((alu3*alu3)+acc2);
  }
  temp[lidx1] = acc2;
  barrier(CLK_LOCAL_MEM_FENCE);
...
```

*uops*
```
...
  33 UOps.BARRIER        :                           [31]                             None
  34 UOps.LOAD           : dtypes.float              [2, 3, 33]                       None
  35 UOps.ALU            : dtypes.float              [34, 4]                          BinaryOps.MUL
  36 UOps.DEFINE_ACC     : dtypes.float              [37]                             (0.0, '1_-1', 0)
  37 UOps.RANGE          : dtypes.int                [3, 6]                           (6, 0)
  38 UOps.ALU            : dtypes.int                [14, 37]                         BinaryOps.ADD
  39 UOps.LOAD           : dtypes.float              [1, 38]                          None
  40 UOps.ALU            : dtypes.float              [39, 35]                         BinaryOps.SUB
  41 UOps.ALU            : dtypes.float              [40, 40]                         BinaryOps.MUL
  42 UOps.ALU            : dtypes.float              [41, 36]                         BinaryOps.ADD
  43 UOps.PHI            : dtypes.float              [36, 42]                         None
  44 UOps.ENDRANGE       :                           [37]                             None
  45 UOps.STORE          :                           [2, 10, 43]                      None
  46 UOps.BARRIER        :                           [45]                             None
...
```

both kernels are correct, so I didn't look into it that much

Reply:
hmm - I don't think it should add the MUL to loop even if it's output is correct. In both cases, you should assert that the mul is always outside of the loop.

Do you have a minimal test for when it's uoped in the loop?

Reply:
I did, but I expanded it to show the correctness issues. It *IS* uoped in the loop, but it's being rendered in the range, ex the steps `cstyle.py` takes when calculating a standard deviation:

```
=== render step 13 ========================================
UOps.ENDRANGE       :                           [<UOps.RANGE: 20>]               None
------------------------------------------------------------
  int gidx0 = get_group_id(0); /* 128 */
  int alu0 = (gidx0*64);
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
============================================================
=== render step 14 ======================================== // hits the MUL here
UOps.ALU            : dtypes.float              [<UOps.PHI: 17>, <UOps.CONST: 6>] BinaryOps.MUL
------------------------------------------------------------
  int gidx0 = get_group_id(0); /* 128 */
  int alu0 = (gidx0*64);
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
============================================================
=== render step 15 ========================================
UOps.DEFINE_ACC     : dtypes.float              [<UOps.RANGE: 20>]               (0.0, '1_0', 0)
------------------------------------------------------------
  int gidx0 = get_group_id(0); /* 128 */
  int alu0 = (gidx0*64);
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
  float acc1 = 0.0f;
============================================================
... // some other ops for rendering the loop
=== render step 20 ========================================
UOps.ALU            : dtypes.float              [<UOps.ALU: 13>, <UOps.DEFINE_ACC: 5>] BinaryOps.ADD
------------------------------------------------------------
  int gidx0 = get_group_id(0); /* 128 */
  int alu0 = (gidx0*64);
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
  float acc1 = 0.0f;
  for (int ridx1 = 0; ridx1 < 64; ridx1++) {
    float val1 = data1[alu0+ridx1];
============================================================
=== render step 21 ========================================
UOps.PHI            : dtypes.float              [<UOps.DEFINE_ACC: 5>, <UOps.ALU: 13>] None // actually renders it here
------------------------------------------------------------
  int gidx0 = get_group_id(0); /* 128 */
  int alu0 = (gidx0*64);
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
  float acc1 = 0.0f;
  for (int ridx1 = 0; ridx1 < 64; ridx1++) {
    float val1 = data1[alu0+ridx1];
    acc1 = ((val1-(acc0*0.015625f))+acc1);
============================================================
=== render step 22 ========================================
UOps.ENDRANGE       :                           [<UOps.RANGE: 20>]               None
------------------------------------------------------------
  int gidx0 = get_group_id(0); /* 128 */
  int alu0 = (gidx0*64);
  float acc0 = 0.0f;
  for (int ridx0 = 0; ridx0 < 64; ridx0++) {
    float val0 = data1[alu0+ridx0];
    acc0 = (val0+acc0);
  }
  float acc1 = 0.0f;
  for (int ridx1 = 0; ridx1 < 64; ridx1++) {
    float val1 = data1[alu0+ridx1];
    acc1 = ((val1-(acc0*0.015625f))+acc1);
  }
============================================================
```

Reply:
The ALU for `acc0*0.015625f` should uop outside of the RANGE, It's ok if cstyle just inlines it though

Reply:
we should be good then 👍